### PR TITLE
OADP-218: Allow for BSL name to be defined in DPA CR

### PR DESCRIPTION
**Pull request by shubham-pampattiwar**

### CODE REVIEW

File: controllers/bsl.go [hunk 1]
```diff
@@ -88,12 +88,20 @@ func (r *DPAReconciler) ReconcileBackupStorageLocations(log logr.Logger) (bool,
...
+		// check if BSL name is specified in DPA spec
+		bslName := bslSpec.Name
+		bslMeta := metav1.ObjectMeta{}
+		if len(bslName) > 0 {
+			bslMeta.Name = bslName
+			bslMeta.Namespace = r.NamespacedName.Namespace
+		} else {
+			bslMeta.Name = fmt.Sprintf("%s-%d", r.NamespacedName.Name, i+1)
+			bslMeta.Namespace = r.NamespacedName.Namespace
+		}
```
Comment (kaovilai) [inline controllers/bsl.go:101]:
```suggestion
		bslName := fmt.Sprintf("%s-%d", r.NamespacedName.Name, i+1)
		if bslSpec.Name != "" {
			bslName = bslSpec.Name
		}
```

File: controllers/bsl.go [hunk 2]
```diff
@@ -88,12 +88,20 @@ func (r *DPAReconciler) ReconcileBackupStorageLocations(log logr.Logger) (bool,
...
+
 		bsl := velerov1.BackupStorageLocation{
-			ObjectMeta: metav1.ObjectMeta{
-				// TODO: Use a hash instead of i
-				Name:      fmt.Sprintf("%s-%d", r.NamespacedName.Name, i+1),
-				Namespace: r.NamespacedName.Namespace,
-			},
+			ObjectMeta: bslMeta,
```
Comment (kaovilai) [inline controllers/bsl.go:104]:
```suggestion
			ObjectMeta: metav1.ObjectMeta{
				Name:      bslName,
				Namespace: r.NamespacedName.Namespace,
			},
```

**Review by kaovilai — CHANGES_REQUESTED** [review]

**Review by eemcmullan — APPROVED** [review]

**Review by kaovilai — APPROVED** [review]

**Comment by kaovilai** [comment]
We should probably follow up with some Reconcile unit tests to define expected behaviors on conflicting names etc. Could just be `expectedErr: true` from a failed create call or assert reconcileStatus expected.

**Pull request merged**